File: mpcop/src/main/java/com/mpcopenplatform/controller/ControllerVerticle.java
package com.mpcopenplatform.controller;

import com.mpcopenplatform.controller.myst.MystVerticle;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.http.HttpServer;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.bridge.PermittedOptions;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.handler.TemplateHandler;
import io.vertx.ext.web.handler.sockjs.SockJSBridgeOptions;
import io.vertx.ext.web.handler.sockjs.SockJSHandler;
import io.vertx.ext.web.handler.sockjs.SockJSHandlerOptions;
import io.vertx.ext.web.templ.handlebars.HandlebarsTemplateEngine;

import java.net.SocketException;
import java.net.UnknownHostException;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.logging.Logger;


/**
 * TODO: refactor
 * The {@link ControllerVerticle} carries communication between the UI controller and
 * the backend verticles.
 *
 * @author Kristian Mika
 */
public class ControllerVerticle extends AbstractVerticle {
    public static final String CONTROLLER_ADDRESS = "service.controller";
    public static final String CONTROLLER_REGISTER_ADDRESS = "service.controller-register";
    public static final String CONTROLLER_BRIDGE_PATH = "/mpcop-event-bus";
    public static final String ROOT_PATH = "/index.hbs";
    public static final String MYST_PATH = "/myst.hbs";
    public static final String SMART_ID_RSA_PATH = "/smart-id-rsa.hbs";
    private static final Logger logger = Logger.getLogger(Logger.GLOBAL_LOGGER_NAME);
    private static final int CONTROLLER_PORT = 8080;
    private static final String HANDLER_HBS_PATH = ".+\\.hbs";
    private static final String HANDLER_RESOURCES_PATH = "/resources/*";
    private static final int NOT_FOUND_ERROR_CODE = 404;
    private static final int HEART_BEAT_INTERVAL = 2000;
    static AtomicInteger connectionCounter = new AtomicInteger();
    TemplateHandler hbsTemplateHandler;

    @Override
    public void start() {
        hbsTemplateHandler = TemplateHandler.create(HandlebarsTemplateEngine.create(vertx));

        HttpServer server = vertx.createHttpServer();

        Router router = Router.router(vertx);

        setHandlers(router);

        SockJSHandlerOptions options = new SockJSHandlerOptions().setHeartbeatInterval(HEART_BEAT_INTERVAL);
        SockJSHandler sockJSHandler = SockJSHandler.create(vertx, options);
        SockJSBridgeOptions bo = new SockJSBridgeOptions()
                .addInboundPermitted(new PermittedOptions().setAddress(CONTROLLER_ADDRESS))
                .addInboundPermitted(new PermittedOptions().setAddress(CONTROLLER_REGISTER_ADDRESS))
                .addOutboundPermitted(new PermittedOptions().setAddress(CONTROLLER_REGISTER_ADDRESS));

        vertx.eventBus().consumer(CONTROLLER_ADDRESS, msg -> {
            vertx.eventBus()
                    .request("service." + getProtocol((JsonObject) msg.body()), msg.body())
                    .onSuccess(data -> {
                        logger.info("Received from " + MystVerticle.CONSUMER_ADDRESS + ": " + data.body().toString());
                        msg.reply(data.body().toString());
                    }).onFailure(thrwbl -> {
                logger.info(Messages.REPLY_FAIL_MESSAGE + thrwbl.toString());
                msg.reply(Messages.ERROR_MESSAGE);
            }).onComplete(msgComp -> {
                logger.info(Messages.REPLY_SUCCESS_MESSAGE);
            });
        });

        sockJSHandler.bridge(bo, event -> {

            switch (event.type()) {
                case REGISTER:
                    logNewConnection(event.socket().remoteAddress().toString());

                    break;
                case UNREGISTER:
                    logClosedConnection(event.socket().remoteAddress().toString());

                    break;
            }
            event.complete(true);
        });

        router.route(CONTROLLER_BRIDGE_PATH + "/*").handler(sockJSHandler);

        server.requestHandler(router).listen(CONTROLLER_PORT);
        logger.info("Controller has been successfully deployed and is now running on port " + CONTROLLER_PORT + ".");
        logger.info(getPrivateIpAnnouncement());

    }

    /**
     * Creates an announcement containing the host's private IP address and instructions on accessing
     * the control panel from other computers.
     */
    private String getPrivateIpAnnouncement() {
        String message = "You can access the control panel from computers within this LAN at ";
        try {
            String privateIp = Util.getPrivateIp();
            message += privateIp + ":" + CONTROLLER_PORT + ".";
        } catch (UnknownHostException | SocketException ignored) {
            message += "port " + CONTROLLER_PORT + " of this computer.";
        }
        return message;
    }

    private void logNewConnection(String address) {
        logger.info("New connection from " + address
                + "\nTotal number of connections: " + connectionCounter.incrementAndGet());
    }

    private void logClosedConnection(String address) {
        logger.info("Connection from " + address + " has been terminated."
                + "\nTotal number of connections: " + connectionCounter.decrementAndGet());
    }


    public static String getProtocol(JsonObject message) {
        return message.getString("protocol");
    }

    void setHandlers(Router router) {
        // A handler for serving css and js resources
        router.get(HANDLER_RESOURCES_PATH).handler(context -> {
            String filename = context.request().path().substring(1);
            vertx.fileSystem().exists(filename, b -> {
                if (b.result()) {
                    context.response().sendFile(filename);
                } else {
                    logger.info("An attempt was made to access '" + context.request().path() + "'.");
                    context.fail(NOT_FOUND_ERROR_CODE);
                }
            });
        });

        // Handler for templated hbs files
        router.getWithRegex(HANDLER_HBS_PATH).handler(hbsTemplateHandler);

        router.get("/").handler(ctx -> {
            ctx.reroute(ROOT_PATH);
        });

        router.get("/myst").handler(ctx -> {
            ctx.reroute(MYST_PATH);
        });

        router.get("/smart-id-rsa").handler(ctx -> {
            ctx.reroute(SMART_ID_RSA_PATH);
        });
    }
}

File: mpcop/src/main/kotlin/cz/muni/fi/mpcop/ControllerVerticle.kt
package cz.muni.fi.mpcop


import cz.muni.fi.mpcop.Utils.getUpdatesAddress
import cz.muni.fi.mpcop.myst.MystVerticle
import cz.muni.fi.mpcop.ping.PingVerticle
import cz.muni.fi.mpcop.smpcrsa.SmpcRsaVerticle
import io.vertx.core.AbstractVerticle
import io.vertx.core.eventbus.ReplyException
import io.vertx.core.eventbus.ReplyFailure
import io.vertx.core.json.JsonObject
import io.vertx.ext.bridge.BridgeEventType
import io.vertx.ext.bridge.PermittedOptions
import io.vertx.ext.web.Router
import io.vertx.ext.web.RoutingContext
import io.vertx.ext.web.handler.StaticHandler
import io.vertx.ext.web.handler.sockjs.SockJSBridgeOptions
import io.vertx.ext.web.handler.sockjs.SockJSHandler
import io.vertx.ext.web.handler.sockjs.SockJSHandlerOptions
import io.vertx.kotlin.core.json.get
import java.net.SocketException
import java.util.concurrent.atomic.AtomicInteger
import java.util.logging.Logger


/**
 * The [ControllerVerticle] facilitates the connections from the clients
 * (runs a http server, serves static content, provides sockJsHandler)
 */
class ControllerVerticle : AbstractVerticle() {

    override fun start() {
        val server = vertx.createHttpServer()
        val router: Router = Router.router(vertx)
        val options: SockJSHandlerOptions = SockJSHandlerOptions().setHeartbeatInterval(HEART_BEAT_INTERVAL)
        val sockJSHandler: SockJSHandler = SockJSHandler.create(vertx, options)
        var bo = SockJSBridgeOptions()

        bo = addInboundPermittedAddresses(
            bo,
            listOf(
                CONTROLLER_ADDRESS,
                CONTROLLER_REGISTER_ADDRESS,
                PingVerticle.verticleAddress
            ) + PROTOCOL_ADDRESSES
        )
        bo = addOutboundPermittedAddresses(
            bo, listOf(
                CONTROLLER_REGISTER_ADDRESS
            ) + PROTOCOL_ADDRESSES.map { address -> getUpdatesAddress(address) }
        )

        bo = bo.setReplyTimeout(REPLY_TIMEOUT)

        sockJSHandler.bridge(bo) { event ->
            when (event.type()) {
                // a new client has registered
                BridgeEventType.REGISTER -> {
                    if (event.rawMessage.get<String>(messageAddressKeyName) == CONTROLLER_REGISTER_ADDRESS) {
                        logNewConnection(event.socket().remoteAddress().toString())
                    }
                }
                // a new client has disconnected
                BridgeEventType.UNREGISTER -> {
                    if (event.rawMessage.get<String>(messageAddressKeyName) == CONTROLLER_REGISTER_ADDRESS) {
                        logClosedConnection(event.socket().remoteAddress().toString())
                    }
                }
                // a client has sent a request
                BridgeEventType.SEND -> {
                    setDefaultHeaders(event.rawMessage)
                    event.rawMessage.get<JsonObject>(messageHeadersKeyName)
                        .put(backendIngresTimestampName, System.currentTimeMillis().toString())
                }
                // a verticle has responded to a request
                BridgeEventType.RECEIVE -> {
                    setDefaultHeaders(event.rawMessage)
                    event.rawMessage.get<JsonObject>(messageHeadersKeyName)
                        .put(backendEgressTimestampName, System.currentTimeMillis().toString())
                }
                else -> {
                }
            }
            event.complete(true)
        }

        // for debug purposes only, logs all internal server errors
        router.errorHandler(HttpCodes.SERVER_ERROR) { rc: RoutingContext ->
            logger.severe("An internal error has occurred: ${rc.failure().message}")
            val failure = rc.failure()
            failure?.printStackTrace()
        }

        // reroute all 404s to index.html
        router.errorHandler(
            HttpCodes.NOT_FOUND
        ) { routingContext: RoutingContext ->
            routingContext.response().setStatusCode(HttpCodes.FOUND)
                .putHeader(HTTP_LOCATION_HEADER_NAME, DEFAULT_REDIRECT_PATH).end()
        }

        router.route("$CONTROLLER_BRIDGE_PATH/*").handler(sockJSHandler)

        val staticHandler = StaticHandler.create(staticContentDir)
        router.route("/*").handler(staticHandler)

        server.requestHandler(router).listen(CONTROLLER_PORT)
        logger.info("Controller has been successfully deployed and is now running on port $CONTROLLER_PORT.")
        logger.info(getPrivateIpAnnouncement())

    }

    /**
     * Adds inbound permitted addresses to bridge [options] - address that can be reached from outside
     * the application event bus
     */
    private fun addInboundPermittedAddresses(
        options: SockJSBridgeOptions,
        addresses: List<String>
    ): SockJSBridgeOptions {
        addresses.forEach { address -> options.addInboundPermitted(PermittedOptions().setAddress(address)) }
        return options
    }

    /**
     * Add outbound permitted addresses to bridge [options] - addresses that can reach outside
     * the application event bus
     */
    private fun addOutboundPermittedAddresses(
        options: SockJSBridgeOptions,
        addresses: List<String>
    ): SockJSBridgeOptions {
        addresses.forEach { address -> options.addOutboundPermitted(PermittedOptions().setAddress(address)) }
        return options
    }

    /**
     * Sets the headers to an empty [JsonObject] in case the [rawMessage] does not contain the header field
     */
    private fun setDefaultHeaders(rawMessage: JsonObject): JsonObject {
        if (!rawMessage.containsKey(messageHeadersKeyName)) {
            rawMessage.put(messageHeadersKeyName, JsonObject())
        }
        return rawMessage
    }


    /**
     * Logs a new connection and increments the connection counter
     *
     * @param address of the new connection
     */
    private fun logNewConnection(address: String) {
        logger.info(
            "New connection from " + address
                    + "\nTotal number of connections: " + connectionCounter.incrementAndGet()
        )
    }

    /**
     * Logs a terminated connection and decrements the connection counter
     *
     * @param address of the terminated connection
     */
    private fun logClosedConnection(address: String) {
        logger.info(
            "Connection from " + address + " has been terminated."
                    + "\nTotal number of connections: " + connectionCounter.decrementAndGet()
        )
    }

    /**
     * Attempts to return an error message with hints on how to resolve the issue based on the throwable
     *
     * @param throwable used for deduction
     * @return error message with hints
     */
    private fun getErrMessage(throwable: Throwable): String {
        if (throwable is ReplyException) {
            if (throwable.failureType() == ReplyFailure.NO_HANDLERS) {
                return Messages.NO_HANDLERS_ERROR
            }
            if (throwable.failureType() == ReplyFailure.TIMEOUT) {
                return Messages.TIMEOUT_ERROR
            }
        }
        return (Messages.GENERIC_ERROR_MESSAGE + " " + throwable.toString())
    }

    /**
     * Creates an announcement containing the host's private IP address and instructions on accessing
     * the control panel from other computers.
     */
    private fun getPrivateIpAnnouncement(): String {
        var message = "You can access the control panel from computers within this LAN at "
        message += try {
            val privateIp: String = Utils.getPrivateIp()
            "$privateIp:${CONTROLLER_PORT}."
        } catch (ignored: SocketException) {
            "port $CONTROLLER_PORT of this computer."
        }
        return message
    }

    companion object {
        const val CONTROLLER_ADDRESS = "service.controller"
        const val CONTROLLER_REGISTER_ADDRESS = "service.controller-register"
        const val CONTROLLER_BRIDGE_PATH = "/mpcop-event-bus"
        val PROTOCOL_ADDRESSES = listOf(MystVerticle.CONSUMER_ADDRESS, SmpcRsaVerticle.CONSUMER_ADDRESS)

        const val backendEgressTimestampName = "backend_egress"
        const val backendIngresTimestampName = "backend_ingress"
        const val messageHeadersKeyName = "headers"
        const val messageAddressKeyName = "address"

        const val HTTP_LOCATION_HEADER_NAME = "Location"
        const val DEFAULT_REDIRECT_PATH = "/index.html"

        private const val staticContentDir: String = "./static"
        private val logger = Logger.getLogger(Logger.GLOBAL_LOGGER_NAME)
        const val CONTROLLER_PORT = 8082
        private const val HEART_BEAT_INTERVAL: Long = 2000
        const val REPLY_TIMEOUT: Long = 60000
        var connectionCounter = AtomicInteger()

    }
}
